community/kernel/src/main/java/org/neo4j/kernel/impl/locking/deferred/DeferringLockClient.java

```diff
@@ -0,0 +1,198 @@
+{
+    private final Locks.Client clientDelegate;
+    private final Map<LockUnit,MutableInt> locks = new TreeMap<>();
+    private volatile boolean stopped;
+
+    public DeferringLockClient( Locks.Client clientDelegate )
+    {
+        this.clientDelegate = clientDelegate;
+    }
+
+    @Override
+    public void acquireShared( Locks.ResourceType resourceType, long... resourceIds ) throws AcquireLockTimeoutException
+    {
+        for ( long resourceId : resourceIds )
+        {
+            addLock( resourceType, resourceId, false );
+        }
+    }
+
+    @Override
+    public void acquireExclusive( Locks.ResourceType resourceType, long... resourceIds )
+            throws AcquireLockTimeoutException
+    {
+        for ( long resourceId : resourceIds )
+        {
+            addLock( resourceType, resourceId, true );
+        }
+    }
+
+    @Override
+    public boolean tryExclusiveLock( Locks.ResourceType resourceType, long resourceId )
+    {
+        throw new UnsupportedOperationException( "Should not be needed" );
+    }
+
+    @Override
+    public boolean trySharedLock( Locks.ResourceType resourceType, long resourceId )
+    {
+        throw new UnsupportedOperationException( "Should not be needed" );
+    }
+
+    @Override
+    public void releaseShared( Locks.ResourceType resourceType, long resourceId )
+    {
+        removeLock( resourceType, resourceId, false );
+    }
+
+    @Override
+    public void releaseExclusive( Locks.ResourceType resourceType, long resourceId )
+    {
+        removeLock( resourceType, resourceId, true );
+    }
+
+    void acquireDeferredLocks()
+    {
+        long[] current = new long[10];
+        int cursor = 0;
+        Locks.ResourceType currentType = null;
+        boolean currentExclusive = false;
+        for ( LockUnit lockUnit : locks.keySet() )
+        {
+            // TODO perhaps also add a condition which sends batches over a certain size threshold
+            if ( currentType == null ||
+                 (currentType.typeId() != lockUnit.resourceType().typeId() ||
+                  currentExclusive != lockUnit.isExclusive()) )
+            {
+                // New type, i.e. flush the current array down to delegate in one call
+                flushLocks( current, cursor, currentType, currentExclusive );
+
+                cursor = 0;
+                currentType = lockUnit.resourceType();
+                currentExclusive = lockUnit.isExclusive();
+            }
+
+            // Queue into current batch
+            if ( cursor == current.length )
+            {
+                current = Arrays.copyOf( current, cursor * 2 );
+            }
+            current[cursor++] = lockUnit.resourceId();
+        }
+        flushLocks( current, cursor, currentType, currentExclusive );
+    }
+
+    private void flushLocks( long[] current, int cursor, Locks.ResourceType currentType, boolean exclusive )
+    {
+        assertNotStopped();
+
+        if ( cursor > 0 )
```

Comment:
Then we will have to make this check in two places. Having it inside the method avoids the duplication.